### DISCUSSION TITLE
JSONB and CompatibleType

### DISCUSSION THREAD

**Discussion by lukesneeringer** [diesel-rs/diesel#3336] [Q&A] [answered]
I'm unable to get a custom type (that worked in 1.4) working in 2.0.

## What I have:

I have a type that corresponds 1:1 with a single table in my database. It derives `Queryable`, although I'm using a custom `SelectStatement` with fields specified. One of the columns is a PostgreSQL `JSONB` field, and that's reflected in the Diesel schema:

```rs
diesel::table! {
  foo_table (id) {
    id -> Varchar,  // UUID
    data -> Jsonb,
  }
}
```

My Rust struct is using a custom struct, because I know what fields to expect. The custom struct has an `impl FromSql<Jsonb, Pg>`, which was how this worked back in 1.4.

```rs
#[derive(Debug, Queryable)]
pub struct FooRow {
  id: String,
  data: MyCustomStruct,
}

impl FromSql<Jsonb, Pg> for FooRow {
  fn from_sql(pg_value: PgValue) -> Result<Self> {
    let value: serde_json::Value = FromSql::<Jsonb, Pg>::from_sql(pg_value)?;
    Ok(Self::from(value))
  }
}
```

I can't get it to play nice with Diesel 2.0. I get an error saying that the `diesel::query_dsl::load_dsl::private::CompatibleType<U, DB>` trait is required and not implemented.

I've tried the following:

- I looked a fair bit at [the code](https://github.com/diesel-rs/diesel/blob/master/diesel/src/query_dsl/load_dsl.rs#L211) and it sure seems like my `FromSql` implementation should be sufficient, but obviously it isn't.
  - I thought the issue might be that my `FromSql` was specific to PostgreSQL and not `Backend`, but it seems to be difficult (maybe even impossible) to implement this for any backend since JSONB is a pg specific type.
  - I tried to determine if `Jsonb` implements `SingleValue`, thinking that could be the problem. I wasn't able to figure that out either way.
- I tried just putting a `serde_json::Value` in my struct to see what would happen. Obviously I got tons of errors in my own code (which is expected with such a type change), but I still saw the compile error about the trait not being implemented.
- I _have not_ tried the "opt into breaking changes" glass-break where I could implement the trait myself, because obviously I want to avoid that. I could do it temporarily if it was a problem in Diesel rather than a problem with my own code.
- I looked around for a test or example that showed the use of a Jsonb value. I saw the tests that just prove that the basic serialization / deserialization works but I couldn't find an exemplar of a Jsonb in an actual schema. (That doesn't mean it's not there, of course, just that I failed to locate it).

I'm stuck. This is my only error before I'm fully migrated. I'd really appreciate it if anyone has helpful advice.

**Comment by weiznich** [comment]
I cannot reproduce this behaviour. For reference the following code worked fine for me:

```rust

diesel::table! {
  foo_table (id) {
    id -> Varchar,  // UUID
    data -> Jsonb,
  }
}

#[derive(Debug, Queryable)]
pub struct FooRow {
  id: String,
  data: MyCustomStruct,
}

#[derive(serde::Deserialize, Debug, FromSqlRow)]
pub struct MyCustomStruct {
    a: i32,
}

impl FromSql<Jsonb, Pg> for MyCustomStruct {
  fn from_sql(pg_value: PgValue) -> Result<Self> {
    let value: serde_json::Value = FromSql::<Jsonb, Pg>::from_sql(pg_value)?;
    Ok(serde_json::from_value(value).unwrap())
  }
}

fn foo(conn: &mut PgConnection) {
    let _: Vec<FooRow> = foo_table::table.load(conn).unwrap();
}


```

**Comment by lukesneeringer** [comment, marked as answer]
I found the problem! Needed a new day or something.

## My mistake

I mis-identified _which_ field was giving me a problem. It's actually a `Vec<String>`, so I was barking up the wrong tree. Sorry for sending you on a goose chase!

## Schema

It looks like the `Vec<String>` is trying to map to `diesel::sql_types::Array<diesel::sql_types::Nullable<diesel::sql_types::Text>>`. That turned out to be my problem.

When I re-generated the schema (which I did because I'm using `diesel-derive-enum`), it generated my PostgreSQL array as a `Array<Nullable<Text>>`, rather than an `Array<Text>`. The relevant column in the migration is `TEXT[] NOT NULL`, so I'd argue that's probably a mistake. (Even if it was nullable, I'd've expected `Nullable<Array<Text>>`.)

Anyway, fixing the schema made the problem go away, and now no problems are detected in the workspace!

## Migration

Overall, thanks for great notes on the migration. I think I completed the whole thing (in a semi-large codebase that was doing some non-trivial custom work) in 4-5ish hours. It would have been closer to 2 if not for my mis-diagnosing the problem. Thanks very much!

**Reply by weiznich** [reply]
> Overall, thanks for great notes on the migration. I think I completed the whole thing (in a semi-large codebase that was doing some non-trivial custom work) in 4-5ish hours. It would have been closer to 2 if not for my mis-diagnosing the problem. Thanks very much!

Thanks for this feedback, that sounds great.

> The relevant column in the migration is TEXT[] NOT NULL, so I'd argue that's probably a mistake. (Even if it was nullable, I'd've expected Nullable<Array<Text>>.)

Such columns can contain nullable values as array elements. That can for example happen due to element updates two elements past the last existing entry. We changed the generated type as part of the 2.0 release. See the relevant section of the [migration guide](https://diesel.rs/guides/migration_guide.html#2-0-0-nullability-of-array-elements) for details.

**Reply by lukesneeringer** [reply]
Ah, okay, I missed that when reading the migration guide. I had just opened #3337, but I'll close it.

**Comment by fenollp** [comment]
Hi there! We've experienced the same issue (how to express non-nullable array *items*) and the only work around we could find was to add a constraint on that `text[] not null` column:
```sql
alter table ips add constraint add_nullcheck_on_tags_array check (array_position(tags, null) is null);
```
(Note: this is invalid SQL: `(text not null)[] not null`)

Now this doesn't get picked up by the schema.rs generator so as to generate a `table!` with a field of type `Array<Text>` (instead of `Array<Nullable<Text>>`).

Should I open an issue repeating this and to track work on fixing the schema.rs generator?
Meanwhile we're relying on a `patch_file` entry in our `diesel.toml`.

**Reply by weiznich** [reply]
> Now this doesn't get picked up by the schema.rs generator so as to generate a table! with a field of type Array<Text> (instead of Array<Nullable<Text>>).

This is expected as diesel-cli generally does not parse and interpret check constraints. That would essentially require parsing and  interpreting backend specific SQL, which is something that's out of scope for diesel.

> Should I open an issue repeating this and to track work on fixing the schema.rs generator?

There is no need to open an issue for this, as this is not a bug. This means there is nothing to fix here from our point of view.

If you feel that the current behaviour can be improved: Feel free to open a new discussion thread in the [idea section](https://github.com/diesel-rs/diesel/discussions/categories/ideas). Such a post should contain at least the following information:

* What can be improved
* Why should this be improved
* How can this be implemented
* What alternatives are there

Each of that points should be addressed as detailed as possible.
